lib/archethic_cache/lru.ex

```diff
@@ -0,0 +1,155 @@
+defmodule ArchethicCache.LRU do
+  @moduledoc """
+  A cache that stores the values in an ETS table.
+  There are hooks available to be able to add effects (ex: write to disk).
+
+  It keeps track of the order and bytes in the genserver state.
+  The `bytes_used` are tracked in here because if we just monitor ETS table size, we would not be able to have a disk cache.
+  The `keys` are used to determine the Least Recent Used (first is the most recent used, last is the least recent used).
+
+  We do not store the values directly in ETS but we insert a pair {size, value} instead.
+  Because size can be modified with the hooks
+  (ex: For LRUDisk, we discard the value from the ETS table, but still want to know the size written to disk)
+  """
+
+  @spec start_link(GenServer.name(), non_neg_integer(), keyword()) ::
+          {:ok, binary()} | {:error, term()}
+  def start_link(name, max_bytes, opts \\ []) do
+    GenServer.start_link(__MODULE__, [name, max_bytes, opts], name: name)
+  end
+
+  @spec put(GenServer.server(), term(), term()) :: :ok
+  def put(server, key, value) do
+    GenServer.cast(server, {:put, key, value})
+  end
+
+  @spec get(GenServer.server(), term()) :: nil | term()
+  def get(server, key) do
+    GenServer.call(server, {:get, key})
+  end
+
+  @spec purge(GenServer.server()) :: :ok
+  def purge(server) do
+    GenServer.call(server, :purge)
+  end
+
+  def init([name, max_bytes, opts]) do
+    table = :ets.new(:"aecache_#{name}", [:set, {:read_concurrency, true}])
+
+    {:ok,
+     %{
+       table: table,
+       bytes_max: max_bytes,
+       bytes_used: 0,
+       keys: [],
+       put_fn: Keyword.get(opts, :put_fn, fn _key, value -> value end),
+       get_fn: Keyword.get(opts, :get_fn, fn _key, value -> value end),
+       evict_fn: Keyword.get(opts, :evict_fn, fn _key, _value -> :ok end)
+     }}
+  end
+
+  def handle_call({:get, key}, _from, state = %{table: table, keys: keys, get_fn: get_fn}) do
+    {reply, new_state} =
+      case :ets.lookup(table, key) do
+        [{^key, {_size, value}}] ->
+          {
+            get_fn.(key, value),
+            %{state | keys: keys |> move_front(key)}
+          }
+
+        [] ->
+          {nil, state}
+      end
+
+    {:reply, reply, new_state}
+  end
+
+  def handle_call(:purge, _from, state = %{table: table, evict_fn: evict_fn}) do
+    # we call the evict_fn to be able to clean effects (ex: file written to disk)
+    :ets.tab2list(table)
```

Comment:
You might use `:ets.foldr` to avoid twice list traversals.
This will save you the overhead of constructing a list for all rows in the table first.